test/distributed/org/apache/cassandra/distributed/test/ClientNetworkStopStartTest.java

```diff
@@ -0,0 +1,189 @@
+            AssertUtils.assertRows(qr, QueryResults.builder().row(0, 0).build());
+        }
+    }
+
+    /**
+     * @see <a href="https://issues.apache.org/jira/browse/CASSANDRA-16127">CASSANDRA-16127</a>
+     */
+    @Test
+    public void stopStartNative() throws IOException
+    {
+        try (Cluster cluster = init(Cluster.build(1).withConfig(c -> c.with(Feature.NATIVE_PROTOCOL)).start()))
+        {
+            IInvokableInstance node = cluster.get(1);
+            assertBinaryRunning(node);
+            assertThriftRunning(node);
+            node.nodetoolResult("disablebinary").asserts().success();
+            assertBinaryNotRunning(node);
+            assertThriftRunning(node);
+            node.nodetoolResult("enablebinary").asserts().success();
+            assertBinaryRunning(node);
+            assertThriftRunning(node);
+
+            // now use it to make sure it still works!
+            cluster.schemaChange("CREATE TABLE " + KEYSPACE + ".tbl (pk int, value int, PRIMARY KEY (pk))");
+
+            try (com.datastax.driver.core.Cluster client = com.datastax.driver.core.Cluster.builder().addContactPoints(node.broadcastAddress().getAddress()).build();
+                 Session session = client.connect())
+            {
+                session.execute("INSERT INTO " + KEYSPACE + ".tbl (pk, value) VALUES (?, ?)", 0, 0);
+            }
+
+            SimpleQueryResult qr = cluster.coordinator(1).executeWithResult("SELECT * FROM " + KEYSPACE + ".tbl", ConsistencyLevel.ALL);
+            AssertUtils.assertRows(qr, QueryResults.builder().row(0, 0).build());
+        }
+    }
+
+    private static void assertBinaryRunning(IInvokableInstance node)
+    {
+        assertNodetoolStdout(node, "running", "not running", "statusbinary");
+    }
+
+    private static void assertBinaryNotRunning(IInvokableInstance node)
+    {
+        assertNodetoolStdout(node, "not running", null, "statusbinary");
+    }
+
+    private static void assertThriftRunning(IInvokableInstance node)
+    {
+        assertNodetoolStdout(node, "running", "not running", "statusthrift");
+    }
+
+    private static void assertThriftNotRunning(IInvokableInstance node)
+    {
+        assertNodetoolStdout(node, "not running", null, "statusthrift");
+    }
+
+    private static void assertNodetoolStdout(IInvokableInstance node, String expectedStatus, String notExpected, String... nodetool)
+    {
+        // without CASSANDRA-16057 need this hack
+        PrintStream previousStdout = System.out;
+        try
+        {
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
+            PrintStream stdout = new PrintStream(out, true);
+            System.setOut(stdout);
+
+            node.nodetoolResult(nodetool).asserts().success();
+
+            stdout.flush();
+            String output = out.toString();
+            Assert.assertThat(output, new StringContains(expectedStatus));
```

Comment:
trunk uses a different version of junit and that has `CoreMatchers.containsString` but this one doesn't, so implemented my own.

Reply:
Yep. It is annoying that the other branches other than trunk are still in `junit 4.6`. Should have port https://issues.apache.org/jira/browse/CASSANDRA-13360 to other branches.